### PR TITLE
Optimization for copying objects between buckets

### PR DESCRIPTION
**Pull request by malcolmgreaves**
This PR optimizes copying objects between buckets. tl;dr is that we do an async copy operation on AWS's backend iff we can.

====from log====

If, when calling copyObject(key,other bucket) on a Bucket instnace, the
client's account owner IDs are the same, then we will kick off a copy
job on the AWS backend asynchronously. This will not incurr any network
costs on our side.

Otherwise, we will resort to copying the object contents directly to
this JVM and then send the contents back to AWS.

### CODE REVIEW

File: src/main/scala/com/nitro/iterator/s3/client.scala [hunk 1]
```diff
@@ -103,7 +103,7 @@ class Bucket(val name: String, val client: AmazonS3) {
           listing <- listingFuture
         } yield {
           listing.objectSummaries match {
-            case Nil => None
+            case Nil          => None
```
Comment (malcolmgreaves) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: formatting picked up by scalariform

File: src/main/scala/com/nitro/iterator/s3/client.scala [hunk 2]
```diff
@@ -141,33 +141,39 @@ class Bucket(val name: String, val client: AmazonS3) {
...
       } match {
 
         // Nothing to do - it's a file copy, so a pure side effect
         case Success(_) =>
+          logger info s"Copied $key to bucket ${other.name}"
+
         // Handle non-fatal errors
         case Failure(NonFatal(t)) =>
-          logger error (s"Failed to copy $key from bucket $name to bucket ${other.name}", t)
+          logger error s"Failed to copy $key from bucket $name to bucket ${other.name}:: $t}"
```
Comment (gregsilin) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: I missed this the first time around.  Providing exception as a separate parameter would actually log the exception properly with a stack race.. any reason we are logging with string interpolation?

File: src/main/scala/com/nitro/iterator/s3/client.scala [hunk 3]
```diff
@@ -141,33 +141,39 @@ class Bucket(val name: String, val client: AmazonS3) {
...
-          other.client.putObject(other.name, key, is, new ObjectMetadata())
+        if (client.getS3AccountOwner.getId == other.client.getS3AccountOwner.getId) {
+          // same AWS account, can do backend async copy
+          client.copyObject(name, key, other.name, key)
+
+        } else {
+          // different AWS accounts, must stream the data into this JVM and then send
+          // it off to be copied
+          for (is <- managed(client.getObject(name, key).getObjectContent)) {
+            other.client.putObject(other.name, key, is, new ObjectMetadata())
```
Comment (gregsilin) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: We just completed some work on Platform to make upload/download asynchronous with Amazon's async transfermanager. We dont' have to block the review on that but I'll open a ticket to port the code

Comment (malcolmgreaves) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: Oh ok, that's groovy. That platform stuff is probably better than this (sorta-hacky) stuff. Should we simply close this PR and then just merge the platform stuff?

File: src/main/scala/com/nitro/iterator/s3/client.scala [hunk 4]
```diff
@@ -141,33 +141,39 @@ class Bucket(val name: String, val client: AmazonS3) {
   val -- = deleteObjects _
 
   /**
-   * Copy object
+   * Copy an object from one bucket to this bucket.
+   *
+   * If the buckets are on the same AWS account, the copy will occur
+   * as an asynchonous computation on the AWS backend. Otherwise, the
+   * data will be copied to this JVM and then sent back to AWS to be
+   * put into the bucket.
```
Comment (pkinsky) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: awesome, comments!

Comment (malcolmgreaves) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: thanks!

File: src/main/scala/com/nitro/iterator/s3/client.scala (hunk 2 above)
Comment (malcolmgreaves) [inline src/main/scala/com/nitro/iterator/s3/client.scala]: Not with respect to that non-fatal error t there. That should be the 2nd argument.

**Comment by malcolmgreaves** [comment]
So I think this stuff should go under the reactive-s3 project since s3/client.scala has been removed.

**Pull request closed without merging**